README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fauxnaif <img src="man/figures/logo.png?raw=TRUE" align="right" height="138" />

<!-- badges: start -->

[![](https://www.r-pkg.org/badges/version/fauxnaif?color=brightgreen)](https://cran.r-project.org/package=fauxnaif)
[![](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![License:
MIT](https://img.shields.io/badge/license-MIT-blueviolet.svg)](https://cran.r-project.org/web/licenses/MIT)
[![R build
status](https://github.com/rossellhayes/fauxnaif/workflows/R-CMD-check/badge.svg)](https://github.com/rossellhayes/fauxnaif/actions)
[![](https://codecov.io/gh/rossellhayes/fauxnaif/branch/master/graph/badge.svg)](https://codecov.io/gh/rossellhayes/fauxnaif)
[![Dependencies](https://tinyverse.netlify.com/badge/fauxnaif)](https://cran.r-project.org/package=fauxnaif)
<!-- badges: end -->

***faux-naïf***
([/ˌfoʊ.naɪˈif/](https://en.wikipedia.org/wiki/Help:IPA/English)): a
person who pretends to be simple or innocent

**fauxnaif**: an R package for simplifying data by pretending values are
`NA`

## Overview

**fauxnaif** provides an extension to `dplyr::na_if()`. Unlike
[**dplyr**](https://github.com/tidyverse/dplyr)’s `na_if()`,
`na_if_in()` allows you to specify multiple values to be replaced with
`NA` using a single function. **fauxnaif** also includes a complementary
function `na_if_not()` to specify values to keep.

## Installation

You can install `fauxnaif` from
[CRAN](https://cran.r-project.org/package=fauxnaif):

``` r
install.packages("fauxanif")
```

Or the development version from
[GitHub](https://github.com/rossellhayes/fauxnaif):

``` r
# install.packages("remotes")
remotes::install_github("rossellhayes/fauxnaif")
```

## Usage

``` r
library(dplyr)
library(fauxnaif)
```

### The basics

Let’s say we want to remove an unwanted negative value from a vector of
numbers

``` r
-1:10
#>  [1] -1  0  1  2  3  4  5  6  7  8  9 10
```

We can replace -1…

… explicitly:

``` r
na_if_in(-1:10, -1)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

… by specifying values to keep:

``` r
na_if_not(-1:10, 0:10)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

… using a formula:

``` r
na_if_in(-1:10, ~ . < 0)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

… or using a function:

``` r
na_if_in(-1:10, min)
#>  [1] NA  0  1  2  3  4  5  6  7  8  9 10
```

### A little more complex

``` r
messy_string <- c("abc", "", "def", "NA", "ghi", 42, "jkl", "NULL", "mno")
```

We can replace unwanted values…

… one at a time:

``` r
na_if_in(messy_string, "")
#> [1] "abc"  NA     "def"  "NA"   "ghi"  "42"   "jkl"  "NULL" "mno"
```

… or all at once:

``` r
na_if_in(messy_string, "", "NA", "NULL", 1:100)
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
na_if_in(messy_string, c("", "NA", "NULL", 1:100))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
na_if_in(messy_string, list("", "NA", "NULL", 1:100))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
```

… or using a clever formula:

``` r
grepl("[a-z]{3,}", messy_string)
#> [1]  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE  TRUE
na_if_not(messy_string, ~ grepl("[a-z]{3,}", .))
#> [1] "abc" NA    "def" NA    "ghi" NA    "jkl" NA    "mno"
```

### With data frames

``` r
faux_census
#> # A tibble: 5 x 4
#>   state    age  income gender                      
#>   <chr>  <dbl>   <dbl> <chr>                       
#> 1 TX        57 9999999 Gender is a social construct
#> 2 Canada    49  149000 Male                        
#> 3 NY       557   90750 f                           
#> 4 LA         2   61000 Male                        
#> 5 TN        64 9999999 M
```

na\_if\_in() is particularly useful inside `dplyr::mutate()`:

``` r
faux_census %>%
 mutate(
   income = na_if_in(income, 9999999),
   age    = na_if_in(age, ~ . < 18, ~ . > 120),
   state  = na_if_not(state, ~ grepl("^[A-Z]{2,}$", .)),
   gender = na_if_in(gender, ~ nchar(.) > 20)
 )
#> # A tibble: 5 x 4
#>   state   age income gender
#>   <chr> <dbl>  <dbl> <chr> 
#> 1 TX       57     NA <NA>  
#> 2 <NA>     49 149000 Male  
#> 3 NY       NA  90750 f     
#> 4 LA       NA  61000 Male  
#> 5 TN       64     NA M
```

Or you can use `dplyr::across()` on data frames:

``` r
faux_census %>%
  mutate(
    across(age, na_if_in, ~ . < 18, ~ . > 120),
    across(state, na_if_not, ~ grepl("^[A-Z]{2,}$", .)),
    across(where(is.character), na_if_in, ~ nchar(.) > 20),
    across(everything(), na_if_in, 9999999)
  )
#> # A tibble: 5 x 4
#>   state   age income gender
#>   <chr> <dbl>  <dbl> <chr> 
#> 1 TX       57     NA <NA>  
#> 2 <NA>     49 149000 Male  
#> 3 NY       NA  90750 f     
#> 4 LA       NA  61000 Male  
#> 5 TN       64     NA M
```

-----

Hex sticker fonts are
[Bodoni\*](https://github.com/indestructible-type/Bodoni) by
[indestructible type\*](https://indestructibletype.com/Home.html) and
[Source Code Pro](https://github.com/adobe-fonts/source-code-pro) by
[Adobe](https://www.adobe.com).

Image adapted from icon made by [Freepik](https://www.freepik.com) from
[flaticon.com](https://www.flaticon.com/free-icon/paper-shredder_1701401).

Please note that **fauxnaif** is released with a [Contributor Code of
Conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/).


NEWS.md
# fauxnaif 0.6.1

* Replace argument `input` with `x`.
* Advance `fauxnaif::na_if()` and scoped `na_if_*()` functions from `deprecate_soft()` to `deprecate_warn()`.
* Only warn `fauxnaif::na_if()` is deprecated if arguments could not be handled by `dplyr::na_if()`.
* Added `pkgdown` site.
* Remove `inst/CITATION`.

# fauxnaif 0.6.0

* No longer produce a warning when no replacements are made.
* Deprecate scoped na_if() functions and suggest replacing them with `dplyr::across()`.
* Deprecate `fauxnaif::na_if()` and suggest replacing it with `na_if_in()`.

# fauxnaif 0.5.6

* Initial CRAN release.
* Scoped na_if_*() functions now call an internal function
  fauxnaif:::scoped_na_if() to reduce repeated code.

# fauxnaif 0.5.5

* Refactored scoped na_if_*() to only message when no replacements are made in
  in entire data frame. Previously, a message would be generated for each column
  in which no replacements were made.

# fauxnaif 0.5.4

* Changed all instances of `sapply()` to `vapply()` or `lapply()` for type
  stability.

# fauxnaif 0.5.3

* Moved package imports from "Depends" to "Imports" for cleaner loading.

# fauxnaif 0.5.2

* Included magrittr pipe.

# fauxnaif 0.5.1

* Rewrote scoped na_if_* functions to use only exported dplyr functions.

# fauxnaif 0.5.0

* Added support for function arguments.
* Complete rewrite of underlying implementation function.

# fauxnaif 0.4.1

* Added unit tests for scoped functions.

# fauxnaif 0.4.0

* Added scoped variants of `na_if()` and `na_if_not()` based on scoped
  `dplyr::mutate_*()` functions.

# fauxnaif 0.3.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat/test-na_if.R
test_that("scalar argument replaces all matching x", {
  expect_equal(na_if_in(0:9, 0), c(NA, 1:9))
  expect_equal(na_if_in(0:9, 1), c(0, NA, 2:9))
  expect_equal(na_if_not(0:9, 0), c(0, rep(NA, 9)))
  expect_equal(na_if_not(0:9, 1), c(NA, 1, rep(NA, 8)))
})

test_that("no changes made if no matches", {
  expect_equal(na_if_in(0:9, 10), 0:9)
  expect_equal(na_if_not(0:9, 0:9), 0:9)
})

test_that("multiple scalar arguments replaces all matching x", {
  expect_equal(na_if_in(0:9, 0, 1), c(NA, NA, 2:9))
  expect_equal(na_if_in(0:9, 0, 9), c(NA, 1:8, NA))
  expect_equal(na_if_not(0:9, 0, 1), c(0, 1, rep(NA, 8)))
  expect_equal(na_if_not(0:9, 0, 9), c(0, rep(NA, 8), 9))
})

test_that("formula argument replaces all matching x", {
  expect_equal(na_if_in(0:9, ~ . < 2), c(NA, NA, 2:9))
  expect_equal(na_if_in(0:9, ~ . >= 8), c(0:7, NA, NA))
  expect_equal(na_if_not(0:9, ~ . < 2), c(0, 1, rep(NA, 8)))
  expect_equal(na_if_not(0:9, ~ . >= 8), c(rep(NA, 8), 8, 9))
})

test_that("function argument replaces all matching x", {
  is.wholenumber <- function(x, tol = .Machine$double.eps^0.5) {
    abs(x - round(x)) < tol
  }

  expect_equal(na_if_in(0:9, min), c(NA, 1:9))
  expect_equal(na_if_in(0:9, max), c(0:8, NA))
  expect_equal(na_if_in(c(0, 0.5, 1), is.wholenumber), c(NA, 0.5, NA))
  expect_equal(na_if_not(0:9, min), c(0, rep(NA, 9)))
  expect_equal(na_if_not(0:9, max), c(rep(NA, 9), 9))
  expect_equal(na_if_not(c(0, 0.5, 1), is.wholenumber), c(0, NA, 1))
})

test_that("non-vector x produces error", {
  expect_error(na_if_in(mean, 0), "Input.*mean")
  expect_error(na_if_in(lm(1 ~ 1), 0), "Input.*lm")
  expect_error(na_if_in(~ . < 0, 0), "Input.*0")
  expect_error(na_if_not(mean, 0), "Input.*mean")
  expect_error(na_if_not(lm(1 ~ 1), 0), "Input.*lm")
  expect_error(na_if_not(~ . < 0, 0), "Input.*0")
})

test_that("coercible non-vector x does not produce error", {
  expect_equal(na_if_in(list(0, 1), 0), list(NA, 1))
  expect_equal(na_if_not(list(0, 1), 0), list(0, NA))
})

test_that("two-sided formula produces warning", {
  expect_warning(na_if_in(0:9, x ~ . < 1), "must be one-sided")
  expect_warning(na_if_not(0:9, x ~ . < 1), "must be one-sided")
})

test_that("non-coercible argument produces warning", {
  expect_warning(na_if_in(0:9, lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_not(0:9, lm(1 ~ 1)), "Argument.*lm")
})

test_that("multiple non-coercible arguments produce multiple warnings", {
  expect_warning(na_if_in(0:9, NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_not(0:9, NULL, lm(1 ~ 1)), "NULL.*lm")
})

test_that("no ... produces warning", {
  expect_warning(na_if_in(0:9), "No values to replace with `NA` specified")
  expect_warning(na_if_not(0:9), "No values to replace with `NA` specified")
})

test_that("na_if() aliases na_if_in()", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))
  expect_equal(na_if(0:9, 0, 1), na_if_in(0:9, 0, 1))
})

test_that("na_if() is deprecated", {
  lifecycle::expect_deprecated(na_if(0:9, 0, 1))
  expect_failure(lifecycle::expect_deprecated(na_if(0:9, 0)))
})


tests/testthat/test-scoped_na_if.R
df <- data.frame(a = 0:9, b = as.numeric(9:0))

test_that("scoped na_if_* errors when dplyr not installed", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  with_mock(
    requireNamespace = function(...) FALSE,
    expect_error(na_if_all(df, 0), "Package `dplyr`"),
    expect_error(na_if_at(df, "a", 0), "Package `dplyr`"),
    expect_error(na_if_if(df, is.integer, 0), "Package `dplyr`"),
    expect_error(na_if_not_all(df, 0), "Package `dplyr`"),
    expect_error(na_if_not_at(df, "a", 0), "Package `dplyr`"),
    expect_error(na_if_not_if(df, is.integer, 0), "Package `dplyr`")
  )
})

test_that("scalar argument replaces all matching x", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  expect_equal(
    na_if_all(df, 0),
    data.frame(a = c(NA, 1:9), b = c(9:1, NA))
  )
  expect_equal(
    na_if_at(df, "a", 0),
    data.frame(a = c(NA, 1:9), b = 9:0)
  )
  expect_equal(
    na_if_if(df, is.integer, 0),
    data.frame(a = c(NA, 1:9), b = 9:0)
  )
  expect_equal(
    na_if_not_all(df, 0),
    data.frame(a = c(0, rep(NA, 9)), b = c(rep(NA, 9), 0))
  )
  expect_equal(
    na_if_not_at(df, "a", 0),
    data.frame(a = c(0, rep(NA, 9)), b = 9:0)
  )
  expect_equal(
    na_if_not_if(df, is.integer, 0),
    data.frame(a = c(0, rep(NA, 9)), b = 9:0)
  )
})

test_that("multiple scalar arguments replaces all matching x", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  df <- data.frame(a = 0:9, b = as.numeric(0:9))

  target <- c(NA, NA, 2:9)
  expect_equal(na_if_all(df, 0, 1), data.frame(a = target, b = target))
  expect_equal(na_if_at(df, "a", 0, 1), data.frame(a = target, b = 0:9))
  expect_equal(na_if_if(df, is.integer, 0, 1), data.frame(a = target, b = 0:9))

  target <- c(NA, 1:8, NA)
  expect_equal(na_if_all(df, 0, 9), data.frame(a = target, b = target))
  expect_equal(na_if_at(df, "a", 0, 9), data.frame(a = target, b = 0:9))
  expect_equal(na_if_if(df, is.integer, 0, 9), data.frame(a = target, b = 0:9))

  target <- c(0, 1, rep(NA, 8))
  expect_equal(na_if_not_all(df, 0, 1), data.frame(a = target, b = target))
  expect_equal(na_if_not_at(df, "a", 0, 1), data.frame(a = target, b = 0:9))
  expect_equal(
    na_if_not_if(df, is.integer, 0, 1), data.frame(a = target, b = 0:9)
  )

  target <- c(0, rep(NA, 8), 9)
  expect_equal(na_if_not_all(df, 0, 9), data.frame(a = target, b = target))
  expect_equal(na_if_not_at(df, "a", 0, 9), data.frame(a = target, b = 0:9))
  expect_equal(
    na_if_not_if(df, is.integer, 0, 9), data.frame(a = target, b = 0:9)
  )
})

test_that("two-sided formula produces warning", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  expect_warning(na_if_all(df, x ~ . < 1), "must be one-sided")
  expect_warning(na_if_at(df, "a", x ~ . < 1), "must be one-sided")
  expect_warning(na_if_if(df, is.integer, x ~ . < 1), "must be one-sided")
  expect_warning(na_if_not_all(df, x ~ . < 1), "must be one-sided")
  expect_warning(na_if_not_at(df, "a", x ~ . < 1), "must be one-sided")
  expect_warning(na_if_not_if(df, is.integer, x ~ . < 1), "must be one-sided")
})

test_that("non-coercible argument produces warning", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  expect_warning(na_if_all(df, lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_at(df, "a", lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_if(df, is.integer, lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_not_all(df, lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_not_at(df, "a", lm(1 ~ 1)), "Argument.*lm")
  expect_warning(na_if_not_if(df, is.integer, lm(1 ~ 1)), "Argument.*lm")
})

test_that("multiple non-coercible arguments produce multiple warnings", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  expect_warning(na_if_all(df, NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_at(df, "a", NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_if(df, is.integer, NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_not_all(df, NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_not_at(df, "a", NULL, lm(1 ~ 1)), "NULL.*lm")
  expect_warning(na_if_not_if(df, is.integer, NULL, lm(1 ~ 1)), "NULL.*lm")
})

test_that("no ... produces warning", {
  withr::local_options(list(lifecycle_verbosity = "quiet"))

  expect_warning(na_if_all(df), "No values")
  expect_warning(na_if_at(df, "a"), "No values")
  expect_warning(na_if_if(df, is.integer), "No values")
  expect_warning(na_if_not_all(df), "No values")
  expect_warning(na_if_not_at(df, "a"), "No values")
  expect_warning(na_if_not_if(df, is.integer), "No values")
})

test_that("deprecated scoped na_if_*()", {
  lifecycle::expect_deprecated(na_if_all(df, 0))
  lifecycle::expect_deprecated(na_if_at(df, "a", 0))
  lifecycle::expect_deprecated(na_if_if(df, is.integer, 0))
  lifecycle::expect_deprecated(na_if_not_all(df, 0))
  lifecycle::expect_deprecated(na_if_not_at(df, "a", 0))
  lifecycle::expect_deprecated(na_if_not_if(df, is.integer, 0))
})


R/faux_na_if.r
#' @importFrom glue glue glue_collapse
#' @importFrom rlang abort as_function as_label caller_env enquo inform
#'     is_atomic is_formula is_function is_list is_logical is_missing warn

faux_na_if <- function(
  x, ..., arguments, arg_names, not = FALSE, scoped = FALSE
) {
  if (!scoped) {
    arguments <- list(...)
    arg_names <- as.list(substitute(list(...)))
  }

  if (is_list(recurse(x))) abort_uncoercible(x)

  if (length(arguments) == 0) {
    warn("No values to replace with `NA` specified.")
    return(x)
  }

  lists <- vapply(arguments, is_list, logical(1))

  arguments[lists] <- lapply(arguments[lists], recurse)

  valid <- vapply(
    arguments,
    function(x) is_atomic(x) | is_function(x) | is_formula(x, lhs = FALSE),
    logical(1)
  )

  eval_replacement <- function(replacement) {
    if (is_atomic(replacement)) return(replacement)

    replacement <- as_function(replacement)(x)

    if (!is_logical(replacement)) return(replacement)

    x[replacement]
  }

  replacements <- recurse(lapply(arguments[valid], eval_replacement))

  if (length(replacements) > 0) {
    if (not) {
      x[!x %in% replacements] <- NA
    } else {
      x[x %in% replacements] <- NA
    }
  }

  two_sided <- vapply(arguments, is_formula, logical(1), lhs = TRUE)

  if (any(two_sided))           warn_two_sided(arg_names[-1][two_sided])
  if (any(!valid & !two_sided)) warn_invalid(arg_names[-1][!valid & !two_sided])

  x
}

scoped_na_if <- function(fun, .tbl, ...) {
  if (!requireNamespace("dplyr")) {
    glue_abort(
      "Package `dplyr` must be installed to use scoped fauxnaif functions.",
      "\n",
      'Try `install.packages("dplyr")`'
    )
  }

  fun    <- get(fun, envir = asNamespace("dplyr"))
  result <- fun(.tbl = .tbl, .funs = faux_na_if, ..., scoped = TRUE)

  result
}


R/scoped_na_if.r
#' Convert values to NA in multiple columns
#'
#' \lifecycle{deprecated}
#'
#' The [dplyr::scoped] variants of [na_if()] and [na_if_not()] can be used
#' directly within pipelines and can modify multiple variables at once.
#'  * `*_all()` affects every variable
#'  * `*_at()` affects variables selected with a character vector or
#'  [dplyr::vars()]
#'  * `*_if()` affects variables selected with a predicate function
#'
#' @param .tbl A `tbl` object
#' @param .predicate A predicate function to be applied to the columns or a
#'   logical vector. The variables for which .predicate is or returns TRUE are
#'   selected. This argument is passed to [rlang::as_function()] and thus
#'   supports quosure-style lambda functions and strings representing function
#'   names.
#' @param .vars A list of columns generated by [dplyr::vars()], a character
#'   vector of column names, a numeric vector of column positions, or NULL.
#' @param ... Values to replace with `NA`, specified as either:
#' \itemize{
#'     \item An object, vector of objects, or list of objects
#'     \item A one-sided formula (see section "Formulas" in [na_if()])
#' }
#' @return A modified data frame. Matched values in selected columns are
#'   replaced with `NA`.
#' @seealso [na_if_in()] and [na_if_not()] operate directly on vectors
#'
#'   [dplyr::mutate_all()], [dplyr::mutate_at()] and [dplyr::mutate_if()] can
#'   apply any function to variables selected in the same way
#'
#' @examples
#' \dontrun{
#' df <- data.frame(a = 0:5, b = 5:0, c = as.numeric(0:5), d = letters[1:6])
#'
#' na_if_all(df, 0)
#' na_if_not_all(df, 0:3, "c")
#'
#' na_if_at(df, c("a", "c"), 0)
#' na_if_not_at(df, c("a", "c"), 0:3)
#'
#' na_if_if(df, is.integer, 0)
#' na_if_not_if(df, is.integer, 0:3)
#' }
#'
#' @name scoped_na_if

NULL

#' @rdname scoped_na_if
#' @export

na_if_all <- function(.tbl, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun       = "mutate_all",
    .tbl      = .tbl,
    arguments = list(...),
    arg_names = as.list(substitute(list(...)))
  )
}

#' @rdname scoped_na_if
#' @export

na_if_not_all <- function(.tbl, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun       = "mutate_all",
    .tbl      = .tbl,
    arguments = list(...),
    arg_names = as.list(substitute(list(...))),
    not       = TRUE
  )
}

#' @rdname scoped_na_if
#' @export

na_if_at <- function(.tbl, .vars, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun       = "mutate_at",
    .tbl      = .tbl,
    .vars     = .vars,
    arguments = list(...),
    arg_names = as.list(substitute(list(...)))
  )
}

#' @rdname scoped_na_if
#' @export

na_if_not_at <- function(.tbl, .vars, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun       = "mutate_at",
    .tbl      = .tbl,
    .vars     = .vars,
    arguments = list(...),
    arg_names = as.list(substitute(list(...))),
    not       = TRUE
  )
}

#' @rdname scoped_na_if
#' @export

na_if_if <- function(.tbl, .predicate, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun        = "mutate_if",
    .tbl       = .tbl,
    .predicate = .predicate,
    arguments  = list(...),
    arg_names  = as.list(substitute(list(...)))
  )
}

#' @rdname scoped_na_if
#' @export

na_if_not_if <- function(.tbl, .predicate, ...) {
  lifecycle::deprecate_warn("0.6.0", format(sys.call()[1]), "dplyr::across()")

  scoped_na_if(
    fun        = "mutate_if",
    .tbl       = .tbl,
    .predicate = .predicate,
    arguments  = list(...),
    arg_names  = as.list(substitute(list(...))),
    not        = TRUE
  )
}



R/error_helpers.r
abort_uncoercible <- function(x) {
  glue_abort(
    'Input must be coercible to a vector. \n* ',
    encodeString(
      as_label(substitute(x, env = parent.frame(2))),
      quote = "`"
    ),
    ' is of uncoercible class "', class(x), '".'
  )
}

warn_invalid <- function(arg_names) {
  glue_warn(
    "Arguments must be atomics, formulas or functions. \n* ",
    glue_list(arg_names),
    " can't be used."
  )
}

warn_two_sided <- function(arg_names) {
  glue_warn(
    "Formulas must be one-sided.\n* ",
    glue_list(arg_names),
    " is a two-sided function."
  )
}

glue_abort <- function(...) {
  abort(glue(...))
}

glue_warn <- function(...) {
  warn(glue(...))
}

glue_list <- function(vector) {
  glue_collapse(tick(vector), sep = ", ", last = " and ")
}

recurse <- function(list) {
  c(list, recursive = TRUE)
}

tick <- function(names) {
  vapply(
    vapply(names, as_label, as_label(1)),
    encodeString,
    character(1),
    quote = "`"
  )
}


R/na_if.r
#' Convert values to NA
#'
#' This is a replacement for [dplyr::na_if()].
#' It is useful if you want to convert annoying values to `NA`.
#' Unlike [dplyr::na_if()], this function allows you to specify multiple values
#' to be replaced with `NA` at the same time.
#'  * `na_if_in()` replaces values that match its arguments with `NA`.
#'  * `na_if_not()` replaces values that *do not* match its arguments with `NA`.
#'
#' @section Formulas:
#' These functions accept one-sided formulas that can evaluate to logical
#' vectors.
#' The input is represented in these conditional statements as "`.`".
#' Valid formulas take the form `~ . < 0`.
#' See examples.
#'
#' @param x Vector to modify
#' @param ... Values to replace with `NA`, specified as either:
#' \itemize{
#'     \item An object, vector of objects, or list of objects
#'     \item A one-sided formula (see section "Formulas")
#' }
#' @return A modified version of `x` with selected values replaced with
#' `NA`.
#'
#' @seealso [dplyr::na_if()] to replace a single value with `NA`.
#'
#'   [dplyr::coalesce()] to replace missing values with a specified value.
#'
#'   [tidyr::replace_na()] to replace `NA` with a value.
#'
#'   [dplyr::recode()] and [dplyr::case_when()] to more generally replace
#'   values.
#'
#' @export
#' @examples
#' -1:10
#' # We can replace -1...
#' # ... explicitly
#' na_if_in(-1:10, -1)
#' # ... by specifying values to keep
#' na_if_not(-1:10, 0:10)
#' # ... using a formula
#' na_if_in(-1:10, ~ . < 0)
#' # ... or using a function
#' na_if_in(-1:10, min)
#'
#' messy_string <- c("abc", "", "def", "NA", "ghi", 42, "jkl", "NULL", "mno")
#' # We can replace unwanted values...
#' # ... one at a time
#' na_if_in(messy_string, "")
#' # ... or all at once
#' na_if_in(messy_string, "", "NA", "NULL", 1:100)
#' na_if_in(messy_string, c("", "NA", "NULL", 1:100))
#' na_if_in(messy_string, list("", "NA", "NULL", 1:100))
#' # ... or using a clever formula
#' grepl("[a-z]{3,}", messy_string)
#' na_if_not(messy_string, ~ grepl("[a-z]{3,}", .))
#'
#' # na_if_in() is particularly useful inside dplyr::mutate
#' library(dplyr)
#' faux_census %>%
#'   mutate(
#'     state = na_if_in(state, "Canada"),
#'     age   = na_if_in(age, ~ . < 18, ~ . > 120)
#'   )
#'
#' # We get a message if our values to replace don't exist
#' na_if_in(-1:10, 11)
#' # And a warning if we use an invalid input...
#' # ... like a two-sided formula
#' na_if_in(-1:10, x ~ . < 0)
#' # ... NULL
#' na_if_in(-1:10, NULL)
#' # ... or nothing at all
#' na_if_in(-1:10)
#'
#' # This function handles vector values differently than dplyr,
#' # and returns a different result with vector replacement values:
#' na_if_in(1:5, 5:1)
#' dplyr::na_if(1:5, 5:1)

na_if_in <- function(x, ...) {
  faux_na_if(x, ...)
}

#' @rdname na_if_in
#' @details # Lifecycle
#'
#' \lifecycle{deprecated} `na_if()` has been deprecated in favor of
#' `na_if_in()` to avoid masking [dplyr::na_if()].
#'
#' @export

na_if <- function(x, ...) {
  if (length(sys.call()) > 3 || !is.atomic(...)) {
    lifecycle::deprecate_warn("0.6.0", "fauxnaif::na_if()", "na_if_in()")
  }

  faux_na_if(x, ...)
}

#' @rdname na_if_in
#' @export

na_if_not <- function(x, ...) {
  faux_na_if(x, ..., not = TRUE)
}
